RecoPixelVertexing/PixelTriplets/plugins/CAHitQuadrupletGenerator.cc

```diff
@@ -0,0 +1,211 @@
+maxChi2(cfg.getParameter<edm::ParameterSet>("maxChi2")),
+fitFastCircle(cfg.getParameter<bool>("fitFastCircle")),
+fitFastCircleChi2Cut(cfg.getParameter<bool>("fitFastCircleChi2Cut")),
+useBendingCorrection(cfg.getParameter<bool>("useBendingCorrection")),
+CAThetaCut(cfg.getParameter<double>("CAThetaCut")),
+CAPhiCut(cfg.getParameter<double>("CAPhiCut"))
+{
+  if (cfg.exists("SeedComparitorPSet"))
+  {
+    edm::ParameterSet comparitorPSet =
+            cfg.getParameter<edm::ParameterSet>("SeedComparitorPSet");
+    std::string comparitorName = comparitorPSet.getParameter<std::string>("ComponentName");
+    if (comparitorName != "none")
+    {
+      theComparitor.reset(SeedComparitorFactory::get()->create(comparitorName, comparitorPSet, iC));
+    }
+  }
+}
+
+CAHitQuadrupletGenerator::~CAHitQuadrupletGenerator()
+{
+}
+
+void CAHitQuadrupletGenerator::hitQuadruplets(
+                                              const TrackingRegion& region, OrderedHitSeeds & result,
+                                              const edm::Event& ev, const edm::EventSetup& es)
+{
+  edm::Handle<SeedingLayerSetsHits> hlayers;
+  ev.getByToken(theSeedingLayerToken, hlayers);
+  const SeedingLayerSetsHits& layers = *hlayers;
+  if (layers.numberOfLayersInSet() != 4)
+    throw cms::Exception("Configuration") << "CAHitQuadrupletsGenerator expects SeedingLayerSetsHits::numberOfLayersInSet() to be 4, got " << layers.numberOfLayersInSet();
+  for (unsigned int j=0; j < layers.size(); j++)
+  {
+    findQuadruplets(region, result, ev, es, layers[j]);
+  }
+
+  theLayerCache.clear();
+}
+
+void
+CAHitQuadrupletGenerator::findQuadruplets (const TrackingRegion& region, OrderedHitSeeds& result,
+                                           const edm::Event& ev, const edm::EventSetup& es,
+                                           const SeedingLayerSetsHits::SeedingLayerSet& fourLayers)
+{
+  if (theComparitor) theComparitor->init (ev, es);
+  HitPairGeneratorFromLayerPair thePairGenerator(0, 1, &theLayerCache);
+
+  std::vector<CACell::CAntuplet> foundQuadruplets;
+
+  std::vector<const HitDoublets*> layersDoublets(3);
+
+  HitDoublets doublets0 =  thePairGenerator.doublets(region, ev, es, fourLayers[0], fourLayers[1] );
+  HitDoublets doublets1 =  thePairGenerator.doublets(region, ev, es, fourLayers[1], fourLayers[2] );
+  HitDoublets doublets2 =  thePairGenerator.doublets(region, ev, es, fourLayers[2], fourLayers[3] );
+
+  layersDoublets[0] = &(doublets0);
+  layersDoublets[1] = &(doublets1);
+  layersDoublets[2] = &(doublets2);
+
+
+  CellularAutomaton<4> ca;
+
+  ca.create_and_connect_cells (layersDoublets, fourLayers, region, CAThetaCut, CAPhiCut);
+
+  ca.evolve();
+
+  ca.find_ntuplets(foundQuadruplets, 4);
+
+  const QuantityDependsPtEval maxChi2Eval = maxChi2.evaluator(es);
+
+
+  // re-used thoughout, need to be vectors because of RZLine interface
+  std::vector<float> bc_r(4), bc_z(4), bc_errZ(4);
+
+  declareDynArray(GlobalPoint, 4, gps);
```

Comment:
Because of interfaces of `RZLine` and `FastCircleFit`, I presume (ok, nowadays they have also the pointer-interface, but I'd leave them to be improved later). Note that the `PixelQuadrupletGenerator` uses exactly the same types.